qiskit/dagcircuit/dagcircuit.py

```diff
@@ -319,6 +319,113 @@ def _add_wire(self, wire):
         else:
             raise DAGCircuitError(f"duplicate wire {wire}")
 
+    def remove_clbits(self, *clbits):
+        """
+        Remove classical bits from the circuit. Only the bits that are idle and not
+        referenced by any classical registers will be removed. Other bits are ignored.
+
+        Raises:
+            DAGCircuitError: a clbit is not a Clbit.
+
+        Returns:
+            set(Clbit): The set of Clbits that were removed.
+        """
+        if any(not isinstance(clbit, Clbit) for clbit in clbits):
+            raise DAGCircuitError("not a Clbit instance.")
+
+        # ignore clbits not in circuit
+        clbits = set(clbits).intersection(self.clbits)
+
+        # ignore clbits referenced by registers
+        register_clbits = {bit for creg in self.cregs.values() for bit in creg}
+        clbits -= register_clbits
+
+        # ignore clbits that are busy
+        busy_clbits = {bit for bit in clbits if not self._is_wire_idle(bit)}
+        clbits -= busy_clbits
```

Comment:
Originally, I did fail loudly for idle and referenced bits, but moved to ignoring instead in 692e0d2d52ba4b09a775d2d7a05e9dde0ca60a66. Previously, the client code was a lot messier, since at least for my use case, it meant I'd need to manually filter out these bits before calling this function. I thought instead it'd be reasonable to have `remove_clbits` return the list of bits that were actually removed, and include a description of this behavior in the doc string. What do you think about giving callers the onus to assert that the return value includes all of the bits they specified if, for their use case, any failed removals should be considered a logic error?

My reasoning for ignoring bits not in the circuit was to sort of make this function idempotent in a sense, i.e. if the bits aren't there, well, that's what we wanted anyways.

> though I'd expect it to work if all the bits in the register are passed

Currently, a call to `remove_cregs` will clean up idle bits if, as a result of the operation, they're no longer referenced. However, `remove_clbits` does not remove registers. If we were to have `remove_clbits` also remove registers, things start to get complex from the caller's perspective IMO. Let's say we want to remove register `A` which has bits `a0-a7`, all of which are idle. However, register `B` has bits `{b0, a0}`, where `b0` is also idle. If `remove_clbits` requires all bits to be specified explicitly, the caller needs to know to pass `a0-a7` and `b0`. At that point, they might as well be dealing in terms of registers. If `remove_clbits` is allowed to remove idle bits not explicitly specified in an effort to clean up registers, the caller can no longer be explicit about which bits they do and do not want to keep.

On the flip side, allowing cleanup of `clbits` by `remove_cregs` works nicely because we can always delete exactly the registers specified by the caller (if underlying `clbits` remain, that's not an issue, so we don't need to delete all registers referencing those bits).

Reply:
I ended up disallowing these things in e59fb8ea6c4462ead5091849c276af42cd85f917 😄.